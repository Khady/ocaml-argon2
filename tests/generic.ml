open Argon2
open Printf

let kind () =
  printf "===== kind =====\n";
  [ D; I; ID ]
  |> List.map (show_kind `Upper)
  |> List.iter (printf "show_kind upper: %s\n");
  [ D; I; ID ]
  |> List.map (show_kind `Lower)
  |> List.iter (printf "show_kind lower: %s\n");
  ()

let show_kind = show_kind `Upper

let hash_len = 32

let t_cost = 2

let m_cost = 65536

let parallelism = 1

let salt = "0000000000000000"

let salt_len = String.length salt

let pwd = "password"

let gen kind =
  let msg fmt = Printf.ksprintf (Printf.printf "%s: %s" (show_kind kind)) fmt in

  let encoded_len =
    encoded_len ~t_cost ~m_cost ~parallelism ~salt_len ~hash_len ~kind
  in
  match
    hash ~t_cost ~m_cost ~parallelism ~pwd ~salt ~kind ~hash_len ~encoded_len
      ~version:VERSION_NUMBER
  with
  | Result.Ok (hash, encoded) -> (
      msg "hash:\n";
      String.iter (fun c -> Printf.printf "%02x" (Char.code c)) hash;
      printf "\n";
      msg "encoded:\n";
      String.iter (fun c -> Printf.printf "%02x" (Char.code c)) encoded;
      printf "\n";
      match verify ~encoded ~pwd ~kind with
      | Result.Ok _ -> msg "verify OK\n"
      | Result.Error e ->
          msg "Error while computing verify: %s\n" (ErrorCodes.message e) )
  | Result.Error e ->
      msg "Error while computing hash: %s\n" (ErrorCodes.message e)

let () =
  kind ();
  gen I;
  gen ID;
  gen D;
  ()
