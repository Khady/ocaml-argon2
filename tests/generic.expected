===== kind =====
show_kind upper: Argon2d
show_kind upper: Argon2i
show_kind upper: Argon2id
show_kind lower: argon2d
show_kind lower: argon2i
show_kind lower: argon2id
Argon2i: hash:
ec6891c09fc1461720e508485da42b2087ca9a708185d9dc890539a33cd2af6f
Argon2i: encoded:
246172676f6e326924763d3139246d3d36353533362c743d322c703d31244d4441774d4441774d4441774d4441774d4441774d412437476952774a2f425268636735516849586151724949664b6d6e434268646e63695155356f7a7a5372323800
Argon2i: verify OK
Argon2id: hash:
36477de377e451ad8a9e38cac8feaa58b5c7e6ee972a4f7ce5fb9df6f5356042
Argon2id: encoded:
246172676f6e32696424763d3139246d3d36353533362c743d322c703d31244d4441774d4441774d4441774d4441774d4441774d41244e6b64393433666b5561324b6e6a6a4b79503671574c5848357536584b6b3938356675643976553159454900
Argon2id: verify OK
Argon2d: hash:
30f5e9f2584e962bba213cc256dcf9ccdce18d8f67a6cbf8e012e0619b6d52d2
Argon2d: encoded:
246172676f6e326424763d3139246d3d36353533362c743d322c703d31244d4441774d4441774d4441774d4441774d4441774d41244d505870386c684f6c69753649547a4356747a357a4e7a686a59396e7073763434424c67595a747455744900
Argon2d: verify OK
